### ISSUE TITLE
Executor parameters change may affect PVF validity and executability retrospectively

### ISSUE THREAD

**Issue by s0me0ne-unkn0wn** [paritytech/polkadot#6876]
Original discussion cited here: https://github.com/paritytech/polkadot/pull/6873#pullrequestreview-1337573839

@s0me0ne-unkn0wn:

What if a PVF is pre-checked in one session and gets enacted in another session? Executor parameters may change in between. There is not much we can do about that, but it's worth considering.

@mrcnski:

I thought about that also. If executor params can affect preparation, then it would invalidate the assumptions documented here. Probably we would have to do pre-checking again when changing any parameter that can affect the runtime.

@s0me0ne-unkn0wn:

We can do that, but what's next? We cannon un-enact a PVF. Consider the situation when it's the first and only PVF for a parachain just onboarded.

**Comment by mrcnski** [comment]
@s0me0ne-unkn0wn Perhaps if an executor param change comes in that could affect preparation, we put the PVF through pre-checking again, and if it fails then we reject the change in params. Not sure how possible this is and sounds complex, but may be worth doing in order to uphold our guarantees around preparation, and prevent disputes there.

**Comment by burdges** [comment]
You could've the validators send some revert limit change votes themselves, maybe the code could be shared with the governance code for doing the limit change somehow?  I don't think it matters if you revert the PVF or the limit change, as either way governance should make choice between the limit change and the existing PVF.

**Comment by bkchr** [comment]
Executor parameter changes will always need to be done after some testing and thinking. I would also expect that we only see parameters to "grow" and not to "shrink". Like the memory for the execution, why should it shrink? If there are any new kind of parameters like a better stack depth algorithm or whatever, we would clearly need to test this with all known PVFs, maybe with all ever existing PVFs to ensure we don't break stuff.

If we really break anything, we need to do what @burdges said and revert our changes.

**Comment by burdges** [comment]
A priori, anytime we shrink then we likely know at least one parachain who already violated the new limits, so I'd hope we already discussed the consequences.

**Comment by mrcnski** [comment]
That is a good point @bkchr. Also, pre-checking is in part intended to guard against adversarial PVFs, whereas adversarial changes in executor params seem unlikely to be enacted.

**Comment by mrcnski** [comment]
Re-posting a comment from @eskimor (from https://github.com/paritytech/polkadot/pull/6873#issuecomment-1469052413):

> We could (not saying we should) don't accept execution environment parameter updates if pre-checking fails for a PVF that was valid before. So changing parameters would also be pre-checked, by re pre-checking all PVFs before enacting them. I actually think this is absolute overkill and does very likely more harm than good, but in that particular case we "could" do something. Realistically, I think we have to accept a small risk and live with it.

**Comment by s0me0ne-unkn0wn** [comment]
> I suppose Wasm stack limit is a stack used by Wasm code, while native stack limit is used by the Wasm engine, our host calls, etc., yes?

Not exactly. The native stack limit is simple, it's the limit of the native stack during the execution of a compiled PVF. It's enforced during the execution, and PVFs do not use host functions afaik, so everything is quite straightforward here.

Wasm stack limit is another beast: it's the limit of the number of values on Wasm value stack. But after the compilation, there's no such thing as a "Wasm value stack" (it's purely abstract) so we enforce this limit through the instrumentation of the original PVF code, and that instrumentation takes place during the preparation phase. So this limit affects preparation and also affects artifacts. We even may have several artifacts for a single PVF if we need to execute it with different executor params from different sessions.

Sidenote: the native stack limit is derived from the wasm stack limit. It was made like that with the assumption that they are proportional (which sounded logical at that moment). Later it was proven they are not. Also, the native stack limit is not deterministic in the general case. So the approach to those limits should be reworked at some point in time anyway.

**Comment by burdges** [comment]
Alright.  We surely have polkadot host functions which PVFs invoke, including signatures verification.  We likely also have host functions used only by polkadot, to which PVFs lack access.

**Comment by s0me0ne-unkn0wn** [comment]
Just an idea: now, after #6934 is merged, new executor parameters are postponed for two sessions and kept in the `PendingConfigs` storage of the `configuration` pallet for that period. Does it make any sense to always pre-check using the pending configuration if different from the current one? Does it make it all click or not? Not sure :thinking:

**Comment by mrcnski** [comment]
Good idea. If we pre-check with the pending configuration, we risk having finality stall if there is an issue with that PVF on the current exec params. 

I think we should pre-check with both configs. And if possible, we should also run preparation again for all PVFs when changing exec params. See https://github.com/paritytech/polkadot/issues/7048.

**Comment by mrcnski** [comment]
Actually I checked [the docs](https://paritytech.github.io/polkadot/book/pvf-prechecking.html) and PVFs are onboarded two sessions after pre-checking. So we should pre-check with the pending config!

> The effects of the PVF accepting depend on the operations requested it:
>
>    All onboardings subscribed to the approved PVF pre-checking process will get scheduled and after passing 2 session boundaries they will be onboarded.
>    All upgrades subscribed to the approved PVF pre-checking process will get scheduled very similarly to the existing process. Upgrades with pre-checking are really the same process that is just delayed by the time required for pre-checking voting. In case of instant approval the mechanism is exactly the same.
